### PR TITLE
Swift: Taint flow improvements

### PR DESCRIPTION
**Pull request by geoffw0**
Taint flow improvements for Swift:
 - expanded the taint flow tests
 - moved the special cases from `UnsafeWebViewFetch.ql` to `TaintTrackingPrivate.qll` so that all queries can benefit.

The next step (in a follow-up PR) will be converting these to models-as-data summaries, which are more concise (but my first attempt isn't working).  With that I should be able to add the required bits to cover most / all of the rest of the test cases quite easily.

### CODE REVIEW

File: swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll [hunk 1]
```diff
@@ -36,11 +36,32 @@ private module Cached {
       nodeTo.asDefinition().(Ssa::WriteDefinition).isInoutDef(e)
     )
     or
+    // allow flow through `try!` and similar constructs
+    // TODO: this should probably be part of DataFlow / TaintTracking?
+    nodeFrom.asExpr() = nodeTo.asExpr().(AnyTryExpr).getSubExpr()
+    or
+    // allow flow through `!`
+    // TODO: this should probably be part of DataFlow / TaintTracking?
+    nodeFrom.asExpr() = nodeTo.asExpr().(ForceValueExpr).getSubExpr()
```
Comment (rdmarsh2) [inline swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll:45]: Should these be in `DataFlow` rather than `TaintTracking`?

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll:45]: Yes, do you know where in `DataFlow` they should be?

Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll:45]: It should be in this predicate: https://github.com/github/codeql/blob/main/swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll#L87

File: swift/ql/test/library-tests/dataflow/taint/data.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+func sink(arg: Data) {}
+func sink2(arg: String) {}
+
+func taintThroughData() {
+	let dataClean = Data("123456".utf8)
+	let dataTainted = Data(source().utf8)
+	let dataTainted2 = Data(dataTainted)
+
+	sink(arg: dataClean)
+	sink(arg: dataTainted) // tainted [NOT DETECTED]
```
Comment (MathiasVP) [inline swift/ql/test/library-tests/dataflow/taint/data.swift:17]: Now that we have InlineExpectationsTest for Swift we should convert out tests to use that framework. No need to do it in this PR, though.

**Comment by geoffw0** [comment]
> I've only added support for MaD sources so far. It shouldn't be difficult to add MaD support for flow steps, though - that work just hasn't been scheduled yet. If we're starting on library modeling we should schedule it soon, though.

Do you know what needs doing for this?  I had a look but couldn't see anything obviously missing (apart from a TODO about return types).

File: swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll [hunk 3]
```diff
@@ -36,11 +36,32 @@ private module Cached {
...
+    or
     // Flow from the computation of the interpolated string literal to the result of the interpolation.
     exists(InterpolatedStringLiteralExpr interpolated |
       nodeTo.asExpr() = interpolated and
       nodeFrom.asExpr() = interpolated.getAppendingExpr()
     )
+    or
+    // allow flow through string concatenation.
+    nodeTo.asExpr().(AddExpr).getAnOperand() = nodeFrom.asExpr()
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll:54]: Good point, I'll address this...

File: swift/ql/test/library-tests/dataflow/taint/data.swift (hunk 2 above)
Comment (geoffw0) [reply swift/ql/test/library-tests/dataflow/taint/data.swift:17]: Yeah, that might be good as a follow-up PR.

File: swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll (hunk 1 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll:45]: Thanks!

**Comment by MathiasVP** [comment]
> Do you know what needs doing for this? I had a look but couldn't see anything obviously missing (apart from a TODO about return types).

We have the [dataflow node](https://github.com/github/codeql/blob/main/swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll#L68) that represents flow through MaD. But that node is not currently being propagated anywhere (i.e., we need to extend the set of `ReturnNode`s and `OutNode`s to include flow out of modeled calls, and we need to extend `ArgumentNode`s and `ParameterNode`s to contain flow into modeled calls).

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll [hunk 4]
```diff
@@ -106,6 +106,12 @@ private module Cached {
       //or
       // step from previous read to Phi node
       localFlowSsaInput(nodeFrom, def, nodeTo.asDefinition())
+      or
+      // flow through `try!` and similar constructs
+      nodeFrom.asExpr() = nodeTo.asExpr().(AnyTryExpr).getSubExpr()
+      or
+      // flow through `!`
+      nodeFrom.asExpr() = nodeTo.asExpr().(ForceValueExpr).getSubExpr()
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:114]: Watch out. This is now a cartesian product with `def`. You should probably move these two cases out of the `exists`.

File: swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll (hunk 1 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll]: Done.

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll (hunk 4 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll]: Well spotted.  Fixed.

I've also attempted to comment a couple of the other cases in this predicate that were lacking comments.

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll [hunk 5]
```diff
@@ -87,25 +87,31 @@
...
~      // use-use flow
~      localSsaFlowStepUseUse(def, nodeFrom, nodeTo)
~      or
~      //localSsaFlowStepUseUse(def, nodeFrom.(PostUpdateNode).getPreUpdateNode(), nodeTo)
       //or
       // step from previous read to Phi node
       localFlowSsaInput(nodeFrom, def, nodeTo.asDefinition())
+      or
+      // flow through `try!` and similar constructs
+      nodeFrom.asExpr() = nodeTo.asExpr().(AnyTryExpr).getSubExpr()
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll]: Fixed.

File: swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/TaintTrackingPrivate.qll]: Done.

**Review by MathiasVP — APPROVED** [review]
LGTM!

File: swift/ql/test/library-tests/dataflow/taint/data.swift (hunk 2 above)
Comment (geoffw0) [reply swift/ql/test/library-tests/dataflow/taint/data.swift:17]: I can make an inline expectation test for the sinks that taint reaches (annotated with where taint came from), but it won't have the same detail (path edges etc) as the query as it is.  Do you think I should keep both queries or prefer just one?

Comment (MathiasVP) [reply swift/ql/test/library-tests/dataflow/taint/data.swift:17]: Yeah, I think there's value in testing the paths as well.

Comment (geoffw0) [reply swift/ql/test/library-tests/dataflow/taint/data.swift:17]: I'm a little reluctant to have three similar but different tests on the same code (when you include `LocalTaint.ql`), but the taint tests are really important.  :+1:

Comment (geoffw0) [reply swift/ql/test/library-tests/dataflow/taint/data.swift:17]: https://github.com/github/codeql/pull/10014

**Pull request merged**